ARROW-14659: [R] Remove warning about factor conversion to string in if_else()

stephhazlitt opened the pull request:
This is my first PR contributing (or an attempt to contribute) to {arrow}.

This PR:

• removes `warn_types` warning that factors are converted to strings, which is no longer true https://github.com/apache/arrow/blob/master/r/R/dplyr-functions.R#L911-L920
• updates the test by removing the warning https://github.com/apache/arrow/blob/master/r/tests/testthat/test-dplyr-funcs-conditional.R#L130 ~however it does not remove the `mutate()` in the test as suggested in the TODO, if removed the test fails?~
• [UPDATE] test includes a reset of the levels of all factor columns to pass, since Arrow if_else() kernel does not preserve unused factor levels (ARROW-14649)

nealrichardson commented:
>  if removed the test fails?

Fails how?

ianmcook commented:
There is a similar test here, in which I solved the same failure by using:
```r
transmute(across(where(is.factor), ~ factor(.x, levels = c(...))))
```
https://github.com/apache/arrow/pull/11272/files?authenticity_token=86nJx4XEwecmmySQatP4CD6%2BYi62xBNQWT4vtOE0iIzJT6wH0ZJM3imNvGEBkfiPii6v0VXSUmag%2B5%2F8ycn0CQ%3D%3D&file-filters%5B%5D=.R#diff-c63a873ac0b560d2ca7229ac1df2628df8396b0cc51fc247b71a2499c492fe3aR317-R320

I think you can use that approach here too!

stephhazlitt commented:
Thanks @ianmcook, I replaced the `as.character` mutate with your approach above and added a comment to reference ARROW-14649. And apologies, I should have added a co-author to that commit.

nealrichardson commented on r/tests/testthat/test-dplyr-funcs-conditional.R:
```diff
@@ -116,18 +116,17 @@ test_that("if_else and ifelse", {
-      mutate(y = as.character(y)),
-    tbl,
-    warning = "Dictionaries .* are currently converted to strings .* in if_else and ifelse"
+      # Arrow coalesce() kernel does not preserve unused factor levels,
```
```suggestion
      # Arrow if_else() kernel does not preserve unused factor levels,
```

ianmcook commented on r/tests/testthat/test-dplyr-funcs-conditional.R:
```diff
@@ -116,18 +116,17 @@ test_that("if_else and ifelse", {
+      # Arrow if_else() kernel does not preserve unused factor levels,
+      # so reset the levels of all the factor columns to make the test pass
+      # (ARROW-14649)
+      transmute(across(where(is.factor), ~ factor(.x, levels = c("a", "b", "c")))),
```
Let's add all the factor levels that are present in `tbl$fct` to avoid creating missing values in the output.
```suggestion
      transmute(across(where(is.factor), ~ factor(.x, levels = c("a", "b", "c", "d", "g", "h", "i", "j")))),
```
